src/gravel/controllers/auth.py

```diff
@@ -0,0 +1,207 @@
+            payload, self._config.jwt_secret,
+            algorithm=self.JWT_ALGORITHM)
+        return encoded_token
+
+    def get_raw_access_token(self, token) -> JWT:
+        raw_token = jwt.decode(token, self._config.jwt_secret,
+                               algorithms=[self.JWT_ALGORITHM])
+        return JWT(**raw_token)
+
+    @staticmethod
+    def set_token_cookie(response: Response, token: str) -> None:
+        response.set_cookie(key=JWTManager.COOKIE_KEY, value=f"Bearer {token}",
+                            httponly=True, samesite="strict")
+
+    @staticmethod
+    def get_token_from_cookie(request: Request) -> Optional[str]:
+        value: Optional[str] = request.cookies.get(JWTManager.COOKIE_KEY)
+        if value is None:
+            return None
+        scheme, token = get_authorization_scheme_param(value)
+        if scheme.lower() == "bearer":
+            return token
+        return None
+
+
+class JWTDenyList:
+    """
+    This list contains JWT tokens that are not allowed to use anymore.
+    E.g. a token is added when a user logs out of the UI. The list will
+    automatically remove expired tokens.
+    """
+
+    def __init__(self, store: KV):
+        self._store: KV = store
+        self._jti_dict: Dict[str, int] = {}
+
+    def _cleanup(self):
+        now = int(datetime.now(timezone.utc).timestamp())
+        self._jti_dict = {jti: exp for jti, exp in self._jti_dict.items() if exp > now}
+
+    async def load(self) -> None:
+        self._jti_dict = {}
+        value = await self._store.get("/auth/jwt_deny_list")
+        if value is not None:
+            self._jti_dict = json.loads(value)
+            self._cleanup()
+
+    async def save(self) -> None:
+        await self._store.put("/auth/jwt_deny_list", json.dumps(self._jti_dict))
+
+    def add(self, token: JWT) -> None:
+        self._jti_dict[token.jti] = token.exp
+
+    def includes(self, token: JWT) -> bool:
+        return token.jti in self._jti_dict
+
+
+class JWTAuthSchema(OAuth2PasswordBearer):
+    def __init__(self):
+        super().__init__(tokenUrl="auth/login")
+
+    async def __call__(self, request: Request) -> Optional[JWT]:  # type: ignore[override]
+        state = request.app.state
+        # Disable authentication as long as the node is not ready.
+        if not state.nodemgr.deployment_state.ready:
+            return None
+        # Get and validate the token.
+        token = JWTManager.get_token_from_cookie(request)
+        if token is None:
+            # Fallback: Try to get it from the headers.
+            token = await super().__call__(request)
+        # Decode token and do the following checks:
+        jwt_mgr = JWTManager(state.gstate.config.options.auth)
+        raw_token: JWT = jwt_mgr.get_raw_access_token(token)
+        # - Is the token revoked?
+        deny_list = JWTDenyList(state.gstate.store)
+        await deny_list.load()
+        if deny_list.includes(raw_token):
+            raise HTTPException(
```

Comment:
Raise a controller specific exception, and translate that to whatever is the appropriate HTTPException in the API controller.

Reply:
That's not possible, an instance of this class is injected in every API endpoint and is processed by FastAPI internally. No way to catch and map it.

Reply:
By moving this to the `api/auth.py` file, we get rid of this annoyance and we raising HTTPExceptions still make sense IMO.